processing/src/main/java/org/apache/druid/query/union/UnionQuery.java

```diff
@@ -0,0 +1,297 @@
+import org.apache.druid.error.DruidException;
+import org.apache.druid.java.util.common.granularity.Granularities;
+import org.apache.druid.java.util.common.granularity.Granularity;
+import org.apache.druid.query.BaseQuery;
+import org.apache.druid.query.DataSource;
+import org.apache.druid.query.Query;
+import org.apache.druid.query.QueryContexts;
+import org.apache.druid.query.QueryRunner;
+import org.apache.druid.query.QuerySegmentWalker;
+import org.apache.druid.query.UnionDataSource;
+import org.apache.druid.query.filter.DimFilter;
+import org.apache.druid.query.planning.DataSourceAnalysis;
+import org.apache.druid.query.spec.QuerySegmentSpec;
+import org.apache.druid.segment.SegmentReference;
+import org.joda.time.DateTimeZone;
+import org.joda.time.Duration;
+import org.joda.time.Interval;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Function;
+
+public class UnionQuery implements Query<UnionResult>
+{
+  @JsonProperty("context")
+  protected final Map<String, Object> context;
+
+  @JsonProperty("queries")
+  protected final List<Query<?>> queries;
+
+  public UnionQuery(List<Query<?>> queries)
+  {
+    this(queries, queries.get(0).getContext());
+  }
+
+  @JsonCreator
+  public UnionQuery(
+      @JsonProperty("queries") List<Query<?>> queries,
+      @JsonProperty("context") Map<String, Object> context)
+  {
+    Preconditions.checkArgument(queries.size() > 1, "union with fewer than 2 queries makes no sense");
+    this.queries = queries;
+    this.context = context;
+  }
+
+  @Override
+  public DataSource getDataSource()
+  {
+    throw DruidException.defensive("This is not supported");
+  }
+
+  @Override
+  public List<DataSource> getDataSources()
+  {
+    List<DataSource> dataSources = new ArrayList<>();
+    for (Query<?> query : queries) {
+      dataSources.add(query.getDataSource());
+    }
+    return dataSources;
+  }
+
+  @Override
+  public boolean hasFilters()
+  {
+    return false;
+  }
+
+  @Override
+  public DimFilter getFilter()
+  {
+    return null;
+  }
+
+  @Override
+  public String getType()
+  {
+    return getClass().getSimpleName();
+  }
+
+  @Override
+  public QueryRunner<UnionResult> getRunner(QuerySegmentWalker walker)
+  {
+    throw DruidException.defensive("Use QueryToolChest to get a Runner");
+  }
+
+  @Override
+  public List<Interval> getIntervals()
+  {
+    return Collections.emptyList();
+  }
+
+  @Override
+  public Duration getDuration()
+  {
+    throw DruidException.defensive("Method supported. [%s]", DruidException.getCurrentMethodName());
+  }
+
+  @Override
+  public Granularity getGranularity()
+  {
+    return Granularities.ALL;
+  }
+
+  @Override
+  public DateTimeZone getTimezone()
+  {
+    throw DruidException.defensive("Method supported. [%s]", DruidException.getCurrentMethodName());
+  }
```

Comment:
I should have not started with this - the exception traces are not always shown; which makes the message less meaningfull. Putting more detail into the exception might help identify the actual issue.

but that shouldn't be addressed individually like I was doing here....it would be better to improve on these things separately.

replaced all these exception creations with a `DruidException#methodNotSupported`